xonsh/proc.py

```diff
@@ -649,6 +652,7 @@ def run(self):
         # kill the process if it is still alive. Happens when piping.
         if proc.poll() is None:
             proc.terminate()
+        events.on_command_finished.fire(cmd=self.command_args, sig=self.previous_signal)
```

Comment:
I don't think this should be on PopenThread.  It makes the most sense for me for this to be in `CommandPipeline.end()`